File: ui/neovim/lua/hemis/events.lua
-- Event client for Hemis backend (Unix socket push notifications)
local config = require("hemis.config")

local M = {}

-- State
M.socket = nil
M.buffer = ""
M.connected = false
M.connecting = false
M.stopped = false
M.reconnect_timer = nil

-- Event handlers (set by init.lua)
M.handlers = {}

-- Paths
local function get_hemis_dir()
  local custom = config.get("hemis_dir")
  if custom then
    return vim.fn.expand(custom)
  end
  return vim.fn.expand("~/.hemis")
end

local function get_events_socket_path()
  return get_hemis_dir() .. "/events.sock"
end

-- Logging
local function log(level, msg)
  local levels = { debug = 1, info = 2, warn = 3, error = 4 }
  local cfg_level = levels[config.get("log_level")] or 3
  if levels[level] >= cfg_level then
    vim.schedule(function()
      vim.notify("[hemis-events] " .. msg, vim.log.levels[level:upper()])
    end)
  end
end

-- Parse JSON-lines from buffer
local function parse_events(buf)
  local events = {}
  local remaining = buf

  while true do
    local newline_pos = remaining:find("\n", 1, true)
    if not newline_pos then
      break
    end

    local line = remaining:sub(1, newline_pos - 1)
    remaining = remaining:sub(newline_pos + 1)

    if #line > 0 then
      local ok, event = pcall(vim.json.decode, line)
      if ok and event and event.type then
        table.insert(events, event)
      else
        log("debug", "Failed to parse event: " .. line)
      end
    end
  end

  return events, remaining
end

-- Handle incoming data
local function on_data(data)
  M.buffer = M.buffer .. data

  local events, remaining = parse_events(M.buffer)
  M.buffer = remaining

  for _, event in ipairs(events) do
    log("debug", "Event: " .. event.type)

    -- Call registered handlers
    local handler = M.handlers[event.type]
    if handler then
      vim.schedule(function()
        handler(event)
      end)
    end

    -- Also call catch-all handler if registered
    if M.handlers["*"] then
      vim.schedule(function()
        M.handlers["*"](event)
      end)
    end
  end
end

-- Schedule reconnect
local function schedule_reconnect()
  if M.stopped or M.reconnect_timer then
    return
  end

  local uv = vim.uv or vim.loop
  M.reconnect_timer = uv.new_timer()
  M.reconnect_timer:start(2000, 0, function()
    M.reconnect_timer:stop()
    M.reconnect_timer:close()
    M.reconnect_timer = nil
    M.connect()
  end)
end

-- Connect to events socket
function M.connect()
  if M.stopped or M.connecting or M.connected then
    return
  end

  local socket_path = get_events_socket_path()

  -- Check if socket exists
  if vim.fn.filereadable(socket_path) == 0 then
    schedule_reconnect()
    return
  end

  M.connecting = true
  log("debug", "Connecting to " .. socket_path)

  local uv = vim.uv or vim.loop
  M.socket = uv.new_pipe(false)

  M.socket:connect(socket_path, function(err)
    M.connecting = false

    if err then
      log("debug", "Connect error: " .. err)
      if M.socket then
        M.socket:close()
        M.socket = nil
      end
      schedule_reconnect()
      return
    end

    M.connected = true
    log("debug", "Connected to events socket")

    -- Start reading
    M.socket:read_start(function(read_err, data)
      if read_err then
        log("debug", "Read error: " .. read_err)
        M.disconnect()
        schedule_reconnect()
        return
      end

      if data then
        on_data(data)
      else
        -- EOF
        log("debug", "Events socket closed")
        M.disconnect()
        schedule_reconnect()
      end
    end)
  end)
end

-- Disconnect
function M.disconnect()
  M.connected = false
  M.buffer = ""

  if M.socket then
    if not M.socket:is_closing() then
      M.socket:close()
    end
    M.socket = nil
  end
end

-- Start event client
function M.start()
  M.stopped = false
  M.connect()
end

-- Stop event client
function M.stop()
  M.stopped = true

  if M.reconnect_timer then
    M.reconnect_timer:stop()
    M.reconnect_timer:close()
    M.reconnect_timer = nil
  end

  M.disconnect()
end

-- Register an event handler
-- event_type: "note-created", "note-position-changed", etc. or "*" for all
function M.on(event_type, handler)
  M.handlers[event_type] = handler
end

-- Check if connected
function M.is_connected()
  return M.connected
end

return M

File: ui/vscode/src/events.ts
import * as net from 'net';
import * as path from 'path';
import * as fs from 'fs';
import { EventEmitter } from 'events';
import { getConfig } from './config';
import { debug } from './debug';

// Event types matching backend/src/events.rs
export interface NoteCreatedEvent {
  type: 'note-created';
  id: string;
  file: string;
  line: number;
  project_root?: string;
}

export interface NoteUpdatedEvent {
  type: 'note-updated';
  id: string;
}

export interface NoteDeletedEvent {
  type: 'note-deleted';
  id: string;
}

export interface NotePositionChangedEvent {
  type: 'note-position-changed';
  id: string;
  file: string;
  old_line: number;
  new_line: number;
  stale: boolean;
}

export interface AiCompleteEvent {
  type: 'ai-complete';
  note_id: string;
  success: boolean;
  provider?: string;
}

export interface IndexCompleteEvent {
  type: 'index-complete';
  project: string;
  files_indexed: number;
}

export interface FileIndexedEvent {
  type: 'file-indexed';
  file: string;
  project: string;
}

export type MnemosEvent =
  | NoteCreatedEvent
  | NoteUpdatedEvent
  | NoteDeletedEvent
  | NotePositionChangedEvent
  | AiCompleteEvent
  | IndexCompleteEvent
  | FileIndexedEvent;

/**
 * Client for receiving push events from the Mnemos backend.
 * Connects to ~/.mnemos/events.sock and emits typed events.
 */
export class EventClient extends EventEmitter {
  private socket: net.Socket | null = null;
  private buffer = '';
  private reconnectTimer: NodeJS.Timeout | null = null;
  private connecting = false;
  private stopped = false;

  private getSocketPath(): string {
    return path.join(getConfig().mnemosDir, 'events.sock');
  }

  /**
   * Start listening for events. Automatically reconnects on disconnect.
   */
  start(): void {
    if (this.socket || this.connecting) {
      return;
    }
    this.stopped = false;
    this.connect();
  }

  /**
   * Stop listening for events.
   */
  stop(): void {
    this.stopped = true;
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
    }
    this.buffer = '';
  }

  private connect(): void {
    if (this.stopped || this.connecting) {
      return;
    }

    const socketPath = this.getSocketPath();
    if (!fs.existsSync(socketPath)) {
      // Socket doesn't exist yet, retry later
      this.scheduleReconnect();
      return;
    }

    this.connecting = true;
    debug('Events: connecting to ' + socketPath);

    const socket = net.createConnection(socketPath);

    socket.on('connect', () => {
      this.connecting = false;
      this.socket = socket;
      debug('Events: connected');
      this.emit('connected');
    });

    socket.on('data', (data: Buffer) => {
      this.onData(data.toString());
    });

    socket.on('close', () => {
      debug('Events: disconnected');
      this.socket = null;
      this.buffer = '';
      this.emit('disconnected');
      this.scheduleReconnect();
    });

    socket.on('error', (err) => {
      debug('Events: error - ' + err.message);
      this.connecting = false;
      // Socket will emit 'close' after error
    });
  }

  private scheduleReconnect(): void {
    if (this.stopped || this.reconnectTimer) {
      return;
    }
    // Reconnect after 2 seconds
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, 2000);
  }

  private onData(data: string): void {
    this.buffer += data;

    // Events are JSON lines (newline-delimited)
    let newlineIndex: number;
    while ((newlineIndex = this.buffer.indexOf('\n')) !== -1) {
      const line = this.buffer.substring(0, newlineIndex);
      this.buffer = this.buffer.substring(newlineIndex + 1);

      if (line.trim()) {
        try {
          const event: MnemosEvent = JSON.parse(line);
          this.handleEvent(event);
        } catch (err) {
          debug('Events: failed to parse - ' + line);
        }
      }
    }
  }

  private handleEvent(event: MnemosEvent): void {
    debug(`Events: ${event.type}`);

    // Emit typed event
    this.emit(event.type, event);

    // Also emit generic 'event' for catch-all handlers
    this.emit('event', event);
  }

  isConnected(): boolean {
    return this.socket !== null;
  }
}

// Singleton instance
let client: EventClient | null = null;

export function getEventClient(): EventClient {
  if (!client) {
    client = new EventClient();
  }
  return client;
}

export function disposeEventClient(): void {
  if (client) {
    client.stop();
    client = null;
  }
}
